src/shared/ui/pickerPrompter.ts

```diff
@@ -256,28 +302,54 @@ export class QuickPickPrompter<T> extends Prompter<T> {
      * previously selected item will remain selected if it still exists after loading.
      *
      * @param items DataQuickPickItems or a promise for said items
+     * @param disableInput Disables the prompter until the items have been loaded, only relevant for async loads (default: true)
      * @returns A promise that is resolved when loading has finished
      */
-    public loadItems(items: Promise<DataQuickPickItem<T>[]> | DataQuickPickItem<T>[]): Promise<void> {
+    public async loadItems(items: ItemLoadTypes<T>, disableInput: boolean = true): Promise<void> {
+        // This code block assumes that callers never try to load items in parallel
+        // For now this okay since we don't have any pickers that require that capability
         const picker = this.quickPick
 
-        if (items instanceof Promise) {
-            picker.busy = true
-            picker.enabled = false
-
-            return items.then(items => {
-                const previousSelected = picker.activeItems
-                picker.items = picker.items.concat(items)
-                this.selectItems(...previousSelected)
-                picker.busy = false
-                picker.enabled = true
-            })
+        if (this.isShowingPlaceholder) {
+            this.clearItems()
+        }
+
+        picker.busy = true
+        picker.enabled = !disableInput
+
+        if (isAsyncIterable(items)) {
+            // Technically AsyncIterators have three types: one for yield, one for return, and one
+            // for parameters to `next`. We only care about the first two, where the yield type will
+            // always be the same as the AsyncIterable type variable, and the second will potentially
+            // be undefined
+            const iterator = items[Symbol.asyncIterator]() as AsyncIterator<
+                DataQuickPickItem<T>[],
+                DataQuickPickItem<T>[] | undefined
+            >
+            // Any caching of the iterator should be handled externally; we will not keep track of
+            // where we left off when the prompt has been hidden
+            let hidden = false
+            const checkHidden = this.quickPick.onDidHide(() => (hidden = true))
+            while (!hidden) {
+                const { value, done } = await iterator.next()
```

Comment:
Should this and the Promise case handle errors with an error node?

Reply:
Yes they should. I actually just haven't gotten around to correctly implementing it since everything that uses `createQuickPick` handles the errors themselves.